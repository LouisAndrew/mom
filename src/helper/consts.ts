export const clickToChatLink = 'https://wa.me/6287782832988';
