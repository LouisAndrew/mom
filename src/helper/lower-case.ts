export const createSlug = (name: string): string =>
    name
        .toLowerCase()
        .split(' ')
        .join('-');
