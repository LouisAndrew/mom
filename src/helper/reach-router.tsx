import React from 'react';
import { LocationProvider } from '@reach/router';

const withReachRouter = (el: any) => <LocationProvider>{el}</LocationProvider>;

export { withReachRouter };
