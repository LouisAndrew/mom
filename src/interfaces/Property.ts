import { FluidObject } from 'gatsby-image';

export type Property = {
    name: string;
    address: string;
    propertyType: 'house' | 'apartment' | 'kavling' | 'home-office';
    saleType: 'sell' | 'rent';

    // optional fields..
    location?: string;
    imgs?: { image: string | FluidObject | FluidObject[]; imgAlt: string }[];
    description?: string;
    isTrending?: boolean;
    geolocation?: { lat: number; lng: number };
    price?: number;
    area?: number;
    roomQty?: number;
    bathQty?: number;
};
