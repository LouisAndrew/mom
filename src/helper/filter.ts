import { intersection, flatten } from 'lodash';

import { Property } from 'interfaces/Property';
import { MinMaxObj } from 'templates/products/products';

type FilterObj = {
    addressFilter: string;
    locationFilters: string[];
    saleTypeFilters: string[];
    propTypeFilters: string[];
    areaFilter: MinMaxObj;
    priceFilter: MinMaxObj;
    properties: Property[];
};

const filterByAddress = (addr: string, properties: Property[]): Property[] =>
    properties.filter(
        property =>
            property.address.toLowerCase().indexOf(addr.toLowerCase()) !== -1
    );

const filterByLocation = (
    locations: string[],
    properties: Property[]
): Property[] =>
    flatten(
        locations.map(location =>
            properties.filter(property => property.location === location)
        )
    );

const filterBySaleType = (
    saleTypes: string[],
    properties: Property[]
): Property[] =>
    flatten(
        saleTypes.map(saleType =>
            properties.filter(property => property.saleType === saleType)
        )
    );

const filterByPropType = (
    propTypes: string[],
    properties: Property[]
): Property[] =>
    flatten(
        propTypes.map(propType =>
            properties.filter(property => property.propertyType === propType)
        )
    );

const filterByArea = (area: MinMaxObj, properties: Property[]): Property[] =>
    properties.filter(
        property =>
            property.area &&
            property.area >= area.min &&
            property.area <= area.max
    );

const filterByPrice = (price: MinMaxObj, properties: Property[]): Property[] =>
    properties.filter(
        property =>
            property.price &&
            property.price >= price.min &&
            property.price <= price.max
    );

/**
 * filters properties given by all filter properties.
 */
export const filter = ({
    addressFilter,
    locationFilters,
    saleTypeFilters,
    propTypeFilters,
    areaFilter,
    priceFilter,
    properties,
}: FilterObj): Promise<Property[]> => {
    const filteredByAddress = filterByAddress(addressFilter, properties);
    // just use the filter if any tags is provided.
    const filteredByLocation =
        locationFilters.length > 0
            ? filterByLocation(locationFilters, properties)
            : properties;
    const filteredBySaleType =
        saleTypeFilters.length > 0
            ? filterBySaleType(saleTypeFilters, properties)
            : properties;
    const filteredByPropType =
        propTypeFilters.length > 0
            ? filterByPropType(propTypeFilters, properties)
            : properties;
    //
    const filteredByArea = filterByArea(areaFilter, properties);
    const filteredByPrice = filterByPrice(priceFilter, properties);

    // returns a promise to provide the ability to use any loader.
    return new Promise<Property[]>((res, rej) => {
        res(
            intersection(
                properties,
                filteredByAddress,
                filteredBySaleType,
                filteredByLocation,
                filteredByPropType,
                filteredByArea,
                filteredByPrice
            )
        );
    });
};
