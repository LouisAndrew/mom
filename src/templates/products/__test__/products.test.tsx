import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';

import Products from '..';
import { Property } from 'interfaces/Property';
import { withReachRouter } from 'helper/reach-router';

describe('All products component', () => {
    const mockProperty: Property[] = [
        {
            name: 'Rumah 1',
            address: 'Jalan rumah 1',
            propertyType: 'house',
            saleType: 'sell',

            imgs: [
                {
                    image:
                        'https://res.cloudinary.com/dsvdffre0/image/upload/v1597137791/webaliser-_TPTXZd9mOo-unsplash_pqtm1r.jpg',
                    imgAlt: 'Beautiful house',
                },
            ],
            description: 'Deskripsi bum bum bum',
            location: 'Penang',
            price: 200,
            area: 125,
        },
        {
            name: 'Rumah 2',
            address: 'Jalan apartment 2',
            propertyType: 'apartment',
            saleType: 'rent',

            imgs: [
                {
                    image:
                        'https://res.cloudinary.com/dsvdffre0/image/upload/v1597137791/jacques-bopp-Hh18POSx5qk-unsplash_lweaht.jpg',
                    imgAlt: 'Beautiful house 2',
                },
            ],
            description: 'Deskripsi bum bum bum',
            location: 'PIK',
            price: 500,
            area: 250,
        },
        {
            name: 'Rumah 10',
            address: 'Jalan ke dufan',
            propertyType: 'home-office',
            saleType: 'rent',

            imgs: [
                {
                    image:
                        'https://res.cloudinary.com/dsvdffre0/image/upload/v1597137792/digital-marketing-agency-ntwrk-g39p1kDjvSY-unsplash_kh6gti.jpg',
                    imgAlt: 'Beautiful house 3',
                },
            ],
            description: 'Deskripsi bum bum bum',
            location: 'jkarta Timur',
            price: 700,
            area: 500,
        },
    ];

    const el: React.ReactElement = withReachRouter(
        <Products properties={mockProperty} />
    ) as React.ReactElement;

    afterEach(cleanup);
    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(el, div);
    });

    it('matches snapshot', () => {
        const tree = renderer.create(el).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
