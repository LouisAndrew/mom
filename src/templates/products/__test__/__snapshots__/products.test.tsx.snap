// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All products component matches snapshot 1`] = `
<div
  className="sc-fzqNqU MkuiH top"
  id="display-top"
>
  <div
    className="sc-AxgMl eyMoUc"
    width={1}
  >
    <div
      className="sc-fzoyTs gfpYqA"
    >
      <span
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <h2
          className="sc-AxirZ iOcOP"
          color={
            Array [
              "dark.2",
            ]
          }
          fontFamily="heading"
          fontSize={
            Array [
              4,
              4,
              5,
              5,
            ]
          }
          width="fit-content"
        >
          Filters 
          0
          <svg
            dangerouslySetInnerHTML={
              Object {
                "__html": "<path d=\\"M8.46 8.29A1 1 0 1 0 7 9.71L9.34 12L7 14.29a1 1 0 0 0 0 1.42a1 1 0 0 0 1.42 0l3-3a1 1 0 0 0 0-1.42zm8.5 3l-3-3a1 1 0 0 0-1.42 1.42l2.3 2.29l-2.3 2.29a1 1 0 0 0 0 1.42a1 1 0 0 0 1.42 0l3-3a1 1 0 0 0 .04-1.42z\\" fill=\\"#3C3639\\"/>",
              }
            }
            focusable={false}
            height="1em"
            preserveAspectRatio="xMidYMid meet"
            style={
              Object {
                "transform": "rotate(360deg)",
                "verticalAlign": "-0.125em",
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          />
        </h2>
      </span>
      <div
        className="sc-fzoNJl bHfWpV"
      />
      <div
        className="sc-fzqARJ ipTxAW"
        display="flex"
        width={1}
      />
    </div>
  </div>
</div>
`;
